### DISCUSSION TITLE
Use long polling in apollo configuration module.

### DISCUSSION THREAD

**Discussion by weixiang1862** [apache/skywalking#11180] [Ideas]
Apollo configuration sync depend on a periodic task(60 S) in oap , when I use agent CDS(ConfigDiscoveryService), I found it usually need some times(agent period + apollo sync period) to receive latest configuration. If we change apollo to [long pooling mode](https://www.apolloconfig.com/#/zh/usage/java-sdk-user-guide?id=_312-%e7%9b%91%e5%90%ac%e9%85%8d%e7%bd%ae%e5%8f%98%e5%8c%96%e4%ba%8b%e4%bb%b6), the time for CDS will nearly equals to agent period, it matches user's imagenation when they set a custom value for agent cds period.

**Comment by wu-sheng** [comment]
I am not familiar with this part.
Do you mean it is using periodically fetching, and you want to change to listening?
@kezhenxu94 I think you know Appollo a lot. Could you take a look?

**Reply by weixiang1862** [reply]
yes, use listening.

**Comment by kezhenxu94** [comment]
Hi @weixiang1862 , thanks for bringing this up, according to [the Apollo doc](https://www.apolloconfig.com/#/zh/usage/java-sdk-user-guide?id=_311-获取默认namespace的配置（application）) (in the previous section right above the link you posted), the `config.getProperty` always get the latest value of the property, why do you say "I found it usually need some times(agent period + apollo sync period) to receive latest configuration", why is it `agent period + apollo sync period`? Unless the doc of Apollo is wrong or there is a bug in Apollo Java client, shouldn't it be the `<= agent period` latency for the agent to receive the changes?

https://github.com/apache/skywalking/blob/80f999f998c7a8277b15b41c32c4fd4c7437a918/oap-server/server-configuration/configuration-apollo/src/main/java/org/apache/skywalking/oap/server/configuration/apollo/ApolloConfigWatcherRegister.java#L62

**Reply by weixiang1862** [reply]
Apollo doc is right, getProperty can get the latest configuration immediately, but it need the task in oap to sync(in every 60seconds). So the time of agent get latest config usually greater the agent default agent CDS period(default is 20 seconds), and the wait time can be 80 seconds in extreme case (agent period + oap sync period)after config changed in apollo.

**Reply by weixiang1862** [reply]
19:57:00  config updated in apollo.
19:58:00  config sync by oap
19:58:20  config discovery by agent.

But it is extrme case, not always like this.

**Reply by kezhenxu94** [reply]
OK, sounds the "latest" value can be an outdated value of the property between two sync tasks are executed, then go ahead to use the long polling mode @weixiang1862

**Reply by weixiang1862** [reply]
Ok, thanks.

**Comment by wu-sheng** [comment]
Updated through https://github.com/apache/skywalking/pull/11186
